From 1bfb013604fbda1cdabf941ee1cef05105c09508 Mon Sep 17 00:00:00 2001
From: Pavel Labath <labath@google.com>
Date: Mon, 9 Jan 2017 09:57:08 +0000
Subject: [PATCH] [lld][cmake] Fix LLVM_LINK_LLVM_DYLIB build

Summary:
Lld's build had a couple of issues which prevented a successfull
LLVM_LINK_LLVM_DYLIB compilation.

- add_llvm_library vs llvm_add_library: One adds a library to libLLVM.so, other
  one doesn't. Lld was using the wrong one, causing symbols to be mupltiply
  defined in things linking to libLLVM.
- confusion when to use LINK_LIBS vs LINK_COMPONENTS in llvm_add_library
- not using LLVM_LINK_COMPONENTS for add_lld_tool

With these fixes lld compiles and it's test suite passes both in
LLVM_LINK_LLVM_DYLIB mode and without it.

Reviewers: ruiu, beanz

Subscribers: llvm-commits, mgorny

Differential Revision: https://reviews.llvm.org/D28397

git-svn-id: https://llvm.org/svn/llvm-project/lld/trunk@291432 91177308-0d34-0410-b5e6-96231b3b80d8
---
 cmake/modules/AddLLD.cmake            |  2 +-
 lib/Config/CMakeLists.txt             |  4 ++--
 lib/Core/CMakeLists.txt               |  4 ++--
 lib/Driver/CMakeLists.txt             |  8 +++++---
 lib/ReaderWriter/CMakeLists.txt       |  6 ++++--
 lib/ReaderWriter/MachO/CMakeLists.txt | 11 +++++++----
 lib/ReaderWriter/YAML/CMakeLists.txt  |  5 ++++-
 tools/lld/CMakeLists.txt              |  5 ++++-
 8 files changed, 29 insertions(+), 16 deletions(-)

diff --git a/cmake/modules/AddLLD.cmake b/cmake/modules/AddLLD.cmake
index 752ca7f..906b295 100644
--- a/cmake/modules/AddLLD.cmake
+++ b/cmake/modules/AddLLD.cmake
@@ -1,5 +1,5 @@
 macro(add_lld_library name)
-  add_llvm_library(${name} ${ARGN})
+  llvm_add_library(${name} ${ARGN})
   set_target_properties(${name} PROPERTIES FOLDER "lld libraries")
 endmacro(add_lld_library)
 
diff --git a/lib/Config/CMakeLists.txt b/lib/Config/CMakeLists.txt
index e971b0b..3e142b6 100644
--- a/lib/Config/CMakeLists.txt
+++ b/lib/Config/CMakeLists.txt
@@ -4,6 +4,6 @@ add_lld_library(lldConfig
   ADDITIONAL_HEADER_DIRS
   ${LLD_INCLUDE_DIR}/lld/Config
 
-  LINK_LIBS
-    LLVMSupport
+  LINK_COMPONENTS
+    Support
   )
diff --git a/lib/Core/CMakeLists.txt b/lib/Core/CMakeLists.txt
index d89ca4a..7f4c47f 100644
--- a/lib/Core/CMakeLists.txt
+++ b/lib/Core/CMakeLists.txt
@@ -12,6 +12,6 @@ add_lld_library(lldCore
   ADDITIONAL_HEADER_DIRS
   ${LLD_INCLUDE_DIR}/lld/Core
 
-  LINK_LIBS
-    LLVMSupport
+  LINK_COMPONENTS
+    Support
   )
diff --git a/lib/Driver/CMakeLists.txt b/lib/Driver/CMakeLists.txt
index 1bd1f21..be75872 100644
--- a/lib/Driver/CMakeLists.txt
+++ b/lib/Driver/CMakeLists.txt
@@ -8,15 +8,17 @@ add_lld_library(lldDriver
   ADDITIONAL_HEADER_DIRS
   ${LLD_INCLUDE_DIR}/lld/Driver
 
+  LINK_COMPONENTS
+    Object
+    Option
+    Support
+
   LINK_LIBS
     lldConfig
     lldMachO
     lldCore
     lldReaderWriter
     lldYAML
-    LLVMObject
-    LLVMOption
-    LLVMSupport
   )
 
 add_dependencies(lldDriver DriverOptionsTableGen)
diff --git a/lib/ReaderWriter/CMakeLists.txt b/lib/ReaderWriter/CMakeLists.txt
index 4408d9c..8751d56 100644
--- a/lib/ReaderWriter/CMakeLists.txt
+++ b/lib/ReaderWriter/CMakeLists.txt
@@ -11,9 +11,11 @@ add_lld_library(lldReaderWriter
   ADDITIONAL_HEADER_DIRS
   ${LLD_INCLUDE_DIR}/lld/ReaderWriter
 
+  LINK_COMPONENTS
+    Object
+    Support
+
   LINK_LIBS
     lldCore
     lldYAML
-    LLVMObject
-    LLVMSupport
   )
diff --git a/lib/ReaderWriter/MachO/CMakeLists.txt b/lib/ReaderWriter/MachO/CMakeLists.txt
index 6a1064d..3b06985 100644
--- a/lib/ReaderWriter/MachO/CMakeLists.txt
+++ b/lib/ReaderWriter/MachO/CMakeLists.txt
@@ -18,11 +18,14 @@ add_lld_library(lldMachO
   StubsPass.cpp
   TLVPass.cpp
   WriterMachO.cpp
+
+  LINK_COMPONENTS
+    Object
+    Support
+
   LINK_LIBS
     lldCore
     lldYAML
-    LLVMObject
-    LLVMSupport
     ${PTHREAD_LIB}
   )
 
diff --git a/lib/ReaderWriter/YAML/CMakeLists.txt b/lib/ReaderWriter/YAML/CMakeLists.txt
index 5c25444..0e63574 100644
--- a/lib/ReaderWriter/YAML/CMakeLists.txt
+++ b/lib/ReaderWriter/YAML/CMakeLists.txt
@@ -1,6 +1,9 @@
 add_lld_library(lldYAML
   ReaderWriterYAML.cpp
+
+  LINK_COMPONENTS
+    Support
+
   LINK_LIBS
     lldCore
-    LLVMSupport
   )
diff --git a/tools/lld/CMakeLists.txt b/tools/lld/CMakeLists.txt
index 0c946a3..2df1069 100644
--- a/tools/lld/CMakeLists.txt
+++ b/tools/lld/CMakeLists.txt
@@ -1,3 +1,7 @@
+set(LLVM_LINK_COMPONENTS
+  Support
+  )
+
 add_lld_tool(lld
   lld.cpp
   )
@@ -6,7 +10,6 @@ target_link_libraries(lld
   lldDriver
   lldCOFF
   lldELF
-  LLVMSupport
   )
 
 install(TARGETS lld
